test/suites/drift/suite_test.go

```diff
@@ -96,6 +97,300 @@ var _ = Describe("Drift", Label("AWS"), func() {
+			}
+
+			By("making the nodes empty")
+			// Delete the deployment to make all nodes empty.
+			env.ExpectDeleted(dep)
+
+			// Drift the nodeclaims
+			By("drift the nodeclaims")
+			nodePool.Spec.Template.Annotations = map[string]string{"test": "annotation"}
+			env.ExpectUpdated(nodePool)
+
+			// Expect that the NodeClaims will all be marked drifted
+			Eventually(func(g Gomega) {
+				nodeClaimList := &corev1beta1.NodeClaimList{}
+				err := env.Client.List(env.Context, nodeClaimList)
+				g.Expect(err).To(Succeed())
+				lo.ForEach(nodeClaimList.Items, func(nc corev1beta1.NodeClaim, _ int) {
+					g.Expect(nc.StatusConditions().GetCondition(corev1beta1.Drifted).IsTrue()).To(BeTrue())
+				})
+			}).Should(Succeed())
+
+			nodes = env.EventuallyExpectTaintedNodeCount("==", 2)
+
+			// Remove the finalizer from each node so that we can terminate
+			for _, node := range nodes {
+				Expect(env.ExpectTestingFinalizerRemoved(node)).To(Succeed())
+			}
+
+			// After the deletion timestamp is set and all pods are drained
+			// the node should be gone
+			env.EventuallyExpectNotFound(nodes[0], nodes[1])
+
+			nodes = env.EventuallyExpectTaintedNodeCount("==", 1)
+			Expect(env.ExpectTestingFinalizerRemoved(nodes[0])).To(Succeed())
+			env.EventuallyExpectNotFound(nodes[0])
+		})
+		It("should respect budgets for non-empty delete drift", func() {
+			nodePool = coretest.ReplaceRequirements(nodePool,
+				v1.NodeSelectorRequirement{
+					Key:      v1beta1.LabelInstanceSize,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{"2xlarge"},
+				},
+			)
+			// We're expecting to create 3 nodes, so we'll expect to see at most 2 nodes deleting at one time.
+			nodePool.Spec.Disruption.Budgets = []corev1beta1.Budget{{
+				Nodes: "50%",
+			}}
+			var numPods int32 = 9
+			dep = coretest.Deployment(coretest.DeploymentOptions{
+				Replicas: numPods,
+				PodOptions: coretest.PodOptions{
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{
+							corev1beta1.DoNotDisruptAnnotationKey: "true",
+						},
+						Labels: map[string]string{"app": "large-app"},
+					},
+					// Each 2xlarge has 8 cpu, so each node should fit no more than 3 pods.
+					ResourceRequirements: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							v1.ResourceCPU: resource.MustParse("2100m"),
+						},
+					},
+				},
+			})
+			selector = labels.SelectorFromSet(dep.Spec.Selector.MatchLabels)
+			env.ExpectCreated(nodeClass, nodePool, dep)
+
+			env.EventuallyExpectCreatedNodeClaimCount("==", 3)
+			env.EventuallyExpectCreatedNodeCount("==", 3)
+			env.EventuallyExpectHealthyPodCount(selector, int(numPods))
+			env.Monitor.Reset() // Reset the monitor so that we can expect a single node to be spun up after drift
+
+			By("scaling down the deployment")
+			// Update the deployment to a third of the replicas.
+			dep.Spec.Replicas = lo.ToPtr[int32](3)
+			env.ExpectUpdated(dep)
+
+			By("spreading the pods to each of the nodes")
```

Comment:
It's not clear to me what scaling-up, back down, and enforcing the spread is testing here. Is this just setup for the rest of the test, i.e. to get 3 2xlarge nodes with 1 pod each, or am I missing something? If it is just setup would it make since to just start with a deployment with a hostname topology spread or larger pod resource requests?

Reply:
Fair question. I chose to spread the pods, because I wanted to first spin up 3 nodes, with more than 1 pod, and then scale it down to a level where they're all non-empty but can reschedule all pods on each other. Larger pod resource requests would make it so that not all 3 pods can schedule onto the nodes, and topology spread would go against the successive delete operations I was trying to achieve.